src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py

```diff
@@ -328,36 +331,56 @@ def cleanRSE(self, rse):
+                    deletedSuccess = 0
+                    for pfnSlice in list(grouper(pfnList, self.msConfig["filesToDeleteSliceSize"])):
+                        try:
                             delResult = ctx.unlink(pfnSlice)
                             # Count all the successfully deleted files (if a deletion was
                             # successful a value of None is put in the delResult list):
-                            deletedSuccess += sum([1 for pfnStatus in delResult if pfnStatus is None])
-
-                        self.logger.debug("RSE: %s, Dir: %s, deletedSuccess: %s",
-                                          rse['name'], dirLfn, deletedSuccess)
-                        rse['counters']['deletedSuccess'] = deletedSuccess
-
-                        # Now clean the whole branch
-                        self.logger.debug("Purging dirEntry: %s:\n", dirPfn)
-                        purgeSuccess = self._purgeTree(ctx, dirPfn)
-                        if not purgeSuccess:
-                            msg = "RSE: %s Failed to purge nonEmpty directory: %s"
-                            self.logger.error(msg, rse['name'], dirPfn)
-                    except Exception as ex:
-                        msg = "Error while cleaning RSE: %s. "
-                        msg += "Will retry in the next cycle. Err: %s"
-                        self.logger.warning(msg, rse['name'], str(ex))
+                            self.logger.debug("RSE: %s, Dir: %s, delResult: %s",
+                                              rse['name'], dirLfn, pformat(delResult))
+                            for gfalErr in delResult:
+                                if gfalErr is None:
+                                    deletedSuccess += 1
+                                elif os.strerror(gfalErr.code) not in rse['counters']['gfalErrors']:
+                                    rse['counters']['gfalErrors'][os.strerror(gfalErr.code)] = 1
+                                else:
+                                    rse['counters']['gfalErrors'][os.strerror(gfalErr.code)] += 1
+                        except Exception as ex:
+                            msg = "Error while cleaning RSE: %s. "
+                            msg += "Will retry in the next cycle. Err: %s"
+                            self.logger.warning(msg, rse['name'], str(ex))
+
+                    self.logger.debug("RSE: %s, Dir: %s, filesDeletedSuccess: %s",
+                                      rse['name'], dirLfn, deletedSuccess)
+                    rse['counters']['filesDeletedSuccess'] += deletedSuccess
+
+                    # Now clean the whole branch
+                    self.logger.debug("Purging dirEntry: %s:\n", dirPfn)
+                    purgeSuccess = self._purgeTree(ctx, dirPfn)
+                    if purgeSuccess:
+                        rse['dirs']['deletedSuccess'].add(dirLfn)
+                        rse['counters']['dirsDeletedSuccess'] = len(rse['dirs']['deletedSuccess'])
```

Comment:
> Does it mean that that pfnSlice is actually a list of file paths within the same directory?

Correct! You have all the list generators working inside one part of the tree only and you have this line L307, which makes it go one dir at a time.   

>  Shouldn't we sum up the dirs deletedSuccess instead of having a direct assignment of that value here?

No, because this is the exact bug we are having right now in production. Since we use the database for state preservation in between the runs we already have a number in the field from previous runs and adding the current successful deletions we risk to make it only increase infinitely and not reflecting the proper state. And since the `rse['counters']['dirsDeletedSuccess']` is a set I assume this to me more accurate because we know exactly what we have removed from sure. And the operation complexity of `len(set)` is still `O(1)`, I think we are not losing anything here, but rather simply staying on the safe side.

Reply:
> Isn't this rse['counters']['dirsDeletedSuccess'] a counter of directories for the whole RSE?

This is True.

>  If so, then I think we should be accumulating directory counters during a cleanup cycle, no?

This is also True.

> Instead of only counting the sub-tree for a given directory (a sub-set of all the directories to be cleaned up).

This is not what is happening here.  
Since we only accumulate in (and never remove from) the `rse['dirs']['deletedSuccess']` set, then every time we refer to its length, we also account for all previous successful deletions on the cost of a simple `O(1)` operation. And because we are working with a set here we avoid adding/counting the same directory twice in the counters. Something that would definitely happen if for some bizarre reason we iterate twice through an already deleted directory (of course we do try to take care of that not to happen but ... one never knows) and then because we consider the already missing directories for success if we blindly count  the current success and add it to the previous counter then we would count the same deletion twice. Hence the bug living in production right now. The counters due to a similar behavior only grow indefinitely.

Reply:
And I see where the confusion may have com - I've made a typo in my previous comment which may have mislead you:
> And since the rse['counters']['dirsDeletedSuccess'] ...

Should have been 
> And since the rse['dirs']['deletedSuccess'] ...

Which is the set accumulating the names of all sucessfully purged branches from the tree. And  not only a sub-tree. 

Apologies for the mistake.